File: Geo_Cortex/Geo_Cortex_Assistant/app/routers/agent.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from typing import Annotated

from app.database import SessionLocal
from app.models.schemas import AgentRequest, AgentResponse, AgentArtifacts
from app.services.master_agent_service import run_master_agent

router = APIRouter(prefix="/agent", tags=["agent"])


def get_db():
    try:
        db = SessionLocal()
        yield db
    finally:
        db.close()


db_dependency = Annotated[Session, Depends(get_db)]


@router.post("/", response_model=AgentResponse)
async def agent_endpoint(request: AgentRequest, db: db_dependency):
    """
    Agentic RAG endpoint (public; no auth).
    The model can run small "processes" (search/count/nearby) before answering.
    """
    answer, trace, occs, artifacts = run_master_agent(db, request.query, max_steps=request.max_steps)
    return AgentResponse(response=answer, tool_trace=trace, occurrences=occs, artifacts=AgentArtifacts(**(artifacts or {})))


# Accept both `/agent` and `/agent/` to avoid 405s when an SPA catch-all route exists.
@router.post("", response_model=AgentResponse, include_in_schema=False)
async def agent_endpoint_noslash(request: AgentRequest, db: db_dependency):
    return await agent_endpoint(request, db)

File: frontend/src/api/agent.ts
import type { AgentRequest, AgentResponse } from "../types/api";
import { httpJson } from "./http";

/**
 * Calls the backend agent endpoint.
 *
 * Backend contract:
 * - Endpoint: POST `${baseUrl}/agent`
 * - Request: { query, max_steps, session_id }
 * - Response: { response, tool_trace, occurrences?, artifacts?, session_id? }
 */
export async function runAgent(baseUrl: string, req: AgentRequest): Promise<AgentResponse> {
  const b = baseUrl.replace(/\/$/, "");
  return await httpJson<AgentResponse>(`${b}/agent/`, {
    method: "POST",
    json: {
      query: req.query,
      max_steps: req.max_steps ?? 3,
      session_id: req.session_id ?? null,
    },
  });
}

